rsocket-core/src/main/java/io/rsocket/RSocketServer.java

```diff
@@ -40,49 +36,51 @@
 
 /**
  * Server side RSocket. Receives {@link Frame}s from a
- * {@link ClientRSocket}
+ * {@link RSocketClient}
  */
-public class ServerRSocket implements RSocket {
+class RSocketServer implements RSocket {
 
     private final DuplexConnection connection;
+    private final RSocket requestHandler;
     private final Consumer<Throwable> errorConsumer;
 
     private final IntObjectHashMap<Subscription> sendingSubscriptions;
     private final IntObjectHashMap<UnicastProcessor<Payload>> channelProcessors;
 
-    private final RSocket requestHandler;
-
-    private volatile Disposable subscribe;
+    private Disposable receiveDisposable;
 
-    public ServerRSocket(DuplexConnection connection, RSocket requestHandler,
-                                boolean clientHonorsLease, Consumer<Throwable> errorConsumer) {
-        this.requestHandler = requestHandler;
+    RSocketServer(DuplexConnection connection,
+                         RSocket requestHandler,
+                         Consumer<Throwable> errorConsumer) {
         this.connection = connection;
-        this.errorConsumer = new KnownErrorFilter(errorConsumer);
+        this.requestHandler = requestHandler;
+        this.errorConsumer = errorConsumer;
         this.sendingSubscriptions = new IntObjectHashMap<>();
         this.channelProcessors = new IntObjectHashMap<>();
+        this.receiveDisposable = connection
+            .receive()
+            .flatMap(this::handleFrame)
+            .doOnError(t -> {
+                errorConsumer.accept(t);
 
-        connection.onClose()
-            .doFinally(signalType -> cleanup())
-            .subscribe();
-        if (requestHandler instanceof LeaseEnforcingSocket) {
-            LeaseEnforcingSocket enforcer = (LeaseEnforcingSocket) requestHandler;
-            enforcer.acceptLeaseSender(lease -> {
-                if (!clientHonorsLease) {
-                    return;
+                Collection<Subscription> values;
+                synchronized (this) {
+                    values = sendingSubscriptions.values();
```

Comment:
IntObjectHashMap::values() looks like a live view of the map.  So I don't trust the concurrency here.